include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+/*
+	When the summing is in exclusive mode - it computes the sum of all the pixels placed
+	on the left and down for a new single texel but it doesn't take sum the main texel itself.
+	In inclusive mode, the texel we start from is taken as well and added to the sum.
+*/
+
+template<bool ExclusiveMode = false>
+class CSummedAreaTableImageFilter : public CMatchedSizeInOutImageFilterCommon, public CSummedAreaTableImageFilterBase<ExclusiveMode>
+{
+	public:
+		virtual ~CSummedAreaTableImageFilter() {}
+
+		class CStateBase : public CMatchedSizeInOutImageFilterCommon::state_type, public CSummedAreaTableImageFilterBase<ExclusiveMode>::CSummStateBase { friend class CSummedAreaTableImageFilter<ExclusiveMode>; };
+		using state_type = CStateBase; //!< full combined state
+
+		static inline bool validate(state_type* state)
+		{
+			if (!CMatchedSizeInOutImageFilterCommon::validate(state))
+				return false;
+
+			if (!CSummedAreaTableImageFilterBase<ExclusiveMode>::validate(state))
+				return false;
+			
+			const ICPUImage::SCreationParams& inParams = state->inImage->getCreationParameters();
+			const ICPUImage::SCreationParams& outParams = state->outImage->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+
+			if (state->scratchMemoryByteSize < state_type::getRequiredScratchByteSize(state->inImage, state->outImage))
+				return false;
+
+			if (asset::getFormatChannelCount(outFormat) != asset::getFormatChannelCount(inFormat))
+				return false;
+
+			if (asset::getFormatClass(inFormat) >= asset::getFormatClass(outFormat)) // TODO in future! Change to a function checking a bit-depth for an each single channel
+				return false;
+
+			return true;
+		}
+
+		static inline bool execute(state_type* state)
+		{
+			if (!validate(state))
+			{
+				state->imageTotalCacheScratch = nullptr;
+				state->imageTotalCacheOutput = nullptr;
+				return false;
+			}
+
+			auto checkFormat = state->inImage->getCreationParameters().format;
+			if (isIntegerFormat(checkFormat))
+				return executeInterprated(state, reinterpret_cast<uint64_t*>(state->scratchMemory));
+			else
+				return executeInterprated(state, reinterpret_cast<double*>(state->scratchMemory));
+		}	
+
+	private:
+
+		template<typename decodeType> //!< double or uint64_t
+		static inline bool executeInterprated(state_type* state, decodeType* scratchMemory)
+		{
+			const asset::E_FORMAT inFormat = state->inImage->getCreationParameters().format;
+			const asset::E_FORMAT outFormat = state->outImage->getCreationParameters().format;
+			const auto inTexelByteSize = asset::getTexelOrBlockBytesize(inFormat);
+			const auto outTexelByteSize = asset::getTexelOrBlockBytesize(outFormat);
+			const auto currentChannelCount = asset::getFormatChannelCount(inFormat);
+			static constexpr auto maxChannels = 4u;
+			const auto outScratchLayerOffset = state->extent.width * state->extent.height * state->extent.depth * currentChannelCount;
```

Comment:
this should be the total extent of your scratch except for the scratch used to return per-layer-max values

Reply:
it isn't valid anymore because I will remove from scratch layers